crates/primitives/src/types/mod.rs

```diff
@@ -0,0 +1,404 @@
+use crate::Block;
+use alloy::primitives::{Address, Bytes, B256, U256};
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, Map};
+
+mod tx;
+pub use tx::{ArchivedTransactionTrace, TransactionTrace, TxL1Msg, TypedTransaction};
+
+/// Block header
+#[derive(
+    rkyv::Archive, rkyv::Serialize, rkyv::Deserialize, Serialize, Deserialize, Default, Debug, Clone,
+)]
+#[archive(check_bytes)]
+#[archive_attr(derive(Debug, Hash, PartialEq, Eq))]
+pub struct BlockHeader {
+    /// block number
+    pub number: U256,
+    /// block hash
+    pub hash: B256,
+    /// timestamp
+    pub timestamp: U256,
+    /// gas limit
+    #[serde(rename = "gasLimit")]
+    pub gas_limit: U256,
+    /// base fee per gas
+    #[serde(rename = "baseFeePerGas")]
+    pub base_fee_per_gas: Option<U256>,
+    /// difficulty
+    pub difficulty: U256,
+    /// mix hash
+    #[serde(rename = "mixHash")]
+    pub mix_hash: Option<B256>,
+}
+
+/// Coinbase
+#[derive(
+    rkyv::Archive, rkyv::Serialize, rkyv::Deserialize, Serialize, Deserialize, Default, Debug, Clone,
+)]
+#[archive(check_bytes)]
+#[archive_attr(derive(Debug, Hash, PartialEq, Eq))]
+pub struct Coinbase {
+    /// address of coinbase
+    pub address: Address,
+}
+
+/// Bytecode trace
+#[derive(
+    rkyv::Archive, rkyv::Serialize, rkyv::Deserialize, Serialize, Deserialize, Default, Debug, Clone,
+)]
+#[archive(check_bytes)]
+#[archive_attr(derive(Debug, Hash, PartialEq, Eq))]
+pub struct BytecodeTrace {
+    /// bytecode
+    pub code: Bytes,
+}
+
+/// storage trace
+#[serde_as]
+#[derive(
+    rkyv::Archive,
+    rkyv::Serialize,
+    rkyv::Deserialize,
+    Serialize,
+    Deserialize,
+    Default,
+    Debug,
+    Clone,
+    Eq,
+    PartialEq,
+)]
+#[archive(check_bytes)]
+#[archive_attr(derive(Debug, Hash, PartialEq, Eq))]
+pub struct StorageTrace {
+    /// root before
+    #[serde(rename = "rootBefore")]
+    pub root_before: B256,
+    /// root after
+    #[serde(rename = "rootAfter")]
+    pub root_after: B256,
+    /// proofs
+    #[serde(rename = "flattenProofs")]
+    #[serde_as(as = "Map<_, _>")]
+    flatten_proofs: Vec<(B256, Bytes)>,
+}
+
+/// Legacy format of block trace
+#[derive(
+    rkyv::Archive, rkyv::Serialize, rkyv::Deserialize, Serialize, Deserialize, Default, Debug, Clone,
+)]
+#[archive(check_bytes)]
+#[archive_attr(derive(Debug, Hash, PartialEq, Eq))]
+pub struct BlockTrace {
```

Comment:
add a ref link to https://github.com/scroll-tech/go-ethereum/blob/develop/core/types/l2trace.go